File: h2/src/main/java/com/dbx/agent/h2/H2Agent.java
package com.dbx.agent.h2;

import com.dbx.agent.AbstractJdbcAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcExecutor;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.ObjectInfo;
import com.dbx.agent.ObjectSource;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;

import java.sql.Connection;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public final class H2Agent extends AbstractJdbcAgent {
    private String databaseName = "";

    @Override
    protected String driverClass() {
        return "org.h2.Driver";
    }

    @Override
    protected String buildJdbcUrl(ConnectParams params) {
        return buildUrl(params);
    }

    @Override
    protected void afterConnect(ConnectParams params, Connection connection) {
        databaseName = params.getDatabase();
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return List.of(new DatabaseInfo(databaseName.isBlank() ? "default" : databaseName));
    }

    @Override
    public List<String> listSchemas() {
        return unchecked(() -> {
            List<String> result = new ArrayList<>();
            try (var stmt = requireConnected().prepareStatement(
                "SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA ORDER BY SCHEMA_NAME"
            );
                 ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    result.add(rs.getString(1));
                }
            }
            return result;
        });
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            List<TableInfo> result = new ArrayList<>();
            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT TABLE_NAME, TABLE_TYPE
                FROM INFORMATION_SCHEMA.TABLES
                WHERE TABLE_SCHEMA = ?
                ORDER BY TABLE_NAME
                """
            )) {
                stmt.setString(1, effectiveSchema);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        String tableType = rs.getString("TABLE_TYPE");
                        if ("BASE TABLE".equals(tableType)) {
                            tableType = "TABLE";
                        }
                        result.add(new TableInfo(rs.getString("TABLE_NAME"), tableType, null));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<ObjectInfo> listObjects(String schema) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            List<ObjectInfo> result = new ArrayList<>();
            for (TableInfo table : listTables(schema)) {
                result.add(new ObjectInfo(table.getName(), table.getTable_type(), schema, table.getComment()));
            }

            try (var stmt = requireConnected().prepareStatement(
                "SELECT ROUTINE_NAME, ROUTINE_TYPE FROM INFORMATION_SCHEMA.ROUTINES WHERE ROUTINE_SCHEMA = ? ORDER BY ROUTINE_NAME"
            )) {
                stmt.setString(1, effectiveSchema);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new ObjectInfo(rs.getString(1), rs.getString(2), schema, null));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public ObjectSource getObjectSource(String schema, String name, String objectType) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            String sql = switch (objectType.toUpperCase(Locale.ROOT)) {
                case "VIEW" -> "SELECT VIEW_DEFINITION FROM INFORMATION_SCHEMA.VIEWS WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?";
                case "FUNCTION", "PROCEDURE" -> "SELECT ROUTINE_DEFINITION FROM INFORMATION_SCHEMA.ROUTINES WHERE ROUTINE_SCHEMA = ? AND ROUTINE_NAME = ?";
                default -> throw new IllegalArgumentException("Unsupported object type: " + objectType);
            };

            String source = "";
            try (var stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, name);
                try (ResultSet rs = stmt.executeQuery()) {
                    if (rs.next()) {
                        String value = rs.getString(1);
                        source = value == null ? "" : value;
                    }
                }
            }
            return new ObjectSource(name, objectType, schema, source);
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            Set<String> primaryKeys = new HashSet<>();
            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT ic.COLUMN_NAME
                FROM INFORMATION_SCHEMA.INDEX_COLUMNS ic
                JOIN INFORMATION_SCHEMA.INDEXES i
                  ON ic.INDEX_SCHEMA = i.INDEX_SCHEMA AND ic.INDEX_NAME = i.INDEX_NAME
                WHERE ic.TABLE_SCHEMA = ? AND ic.TABLE_NAME = ?
                  AND i.INDEX_TYPE_NAME = 'PRIMARY KEY'
                """
            )) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        primaryKeys.add(rs.getString("COLUMN_NAME"));
                    }
                }
            }

            List<ColumnInfo> result = new ArrayList<>();
            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT COLUMN_NAME, DATA_TYPE, IS_NULLABLE, COLUMN_DEFAULT,
                       NUMERIC_PRECISION, NUMERIC_SCALE, CHARACTER_MAXIMUM_LENGTH
                FROM INFORMATION_SCHEMA.COLUMNS
                WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
                ORDER BY ORDINAL_POSITION
                """
            )) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        String columnName = rs.getString("COLUMN_NAME");
                        result.add(new ColumnInfo(
                            columnName,
                            rs.getString("DATA_TYPE"),
                            "YES".equals(rs.getString("IS_NULLABLE")),
                            rs.getString("COLUMN_DEFAULT"),
                            primaryKeys.contains(columnName),
                            null,
                            null,
                            intOrNull(rs, "NUMERIC_PRECISION"),
                            intOrNull(rs, "NUMERIC_SCALE"),
                            intOrNull(rs, "CHARACTER_MAXIMUM_LENGTH")
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            Map<String, List<String>> indexMap = new LinkedHashMap<>();
            Map<String, Boolean> uniqueMap = new HashMap<>();
            Map<String, Boolean> primaryMap = new HashMap<>();
            Map<String, String> typeMap = new HashMap<>();

            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT i.INDEX_NAME, ic.COLUMN_NAME, ic.IS_UNIQUE, i.INDEX_TYPE_NAME
                FROM INFORMATION_SCHEMA.INDEX_COLUMNS ic
                JOIN INFORMATION_SCHEMA.INDEXES i
                  ON ic.INDEX_SCHEMA = i.INDEX_SCHEMA AND ic.INDEX_NAME = i.INDEX_NAME
                WHERE ic.TABLE_SCHEMA = ? AND ic.TABLE_NAME = ?
                ORDER BY i.INDEX_NAME, ic.ORDINAL_POSITION
                """
            )) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        String indexName = rs.getString("INDEX_NAME");
                        String columnName = rs.getString("COLUMN_NAME");
                        String indexType = rs.getString("INDEX_TYPE_NAME");

                        indexMap.computeIfAbsent(indexName, ignored -> new ArrayList<>()).add(columnName);
                        uniqueMap.put(indexName, rs.getBoolean("IS_UNIQUE"));
                        primaryMap.put(indexName, "PRIMARY KEY".equals(indexType));
                        typeMap.put(indexName, indexType == null ? "" : indexType);
                    }
                }
            }

            List<IndexInfo> result = new ArrayList<>();
            for (Map.Entry<String, List<String>> entry : indexMap.entrySet()) {
                String name = entry.getKey();
                result.add(new IndexInfo(
                    name,
                    entry.getValue(),
                    uniqueMap.getOrDefault(name, false),
                    primaryMap.getOrDefault(name, false),
                    null,
                    typeMap.get(name),
                    null,
                    null
                ));
            }
            return result;
        });
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            List<ForeignKeyInfo> result = new ArrayList<>();
            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT FK_NAME, FKCOLUMN_NAME, PKTABLE_NAME, PKCOLUMN_NAME
                FROM INFORMATION_SCHEMA.CROSS_REFERENCES
                WHERE FKTABLE_SCHEMA = ? AND FKTABLE_NAME = ?
                ORDER BY FK_NAME
                """
            )) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new ForeignKeyInfo(
                            rs.getString("FK_NAME"),
                            rs.getString("FKCOLUMN_NAME"),
                            rs.getString("PKTABLE_NAME"),
                            rs.getString("PKCOLUMN_NAME")
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return unchecked(() -> {
            String effectiveSchema = resolveSchema(schema);
            List<TriggerInfo> result = new ArrayList<>();
            try (var stmt = requireConnected().prepareStatement(
                """
                SELECT TRIGGER_NAME, EVENT_MANIPULATION, ACTION_TIMING
                FROM INFORMATION_SCHEMA.TRIGGERS
                WHERE TRIGGER_SCHEMA = ? AND EVENT_OBJECT_TABLE = ?
                ORDER BY TRIGGER_NAME
                """
            )) {
                stmt.setString(1, effectiveSchema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new TriggerInfo(
                            rs.getString("TRIGGER_NAME"),
                            rs.getString("EVENT_MANIPULATION"),
                            rs.getString("ACTION_TIMING")
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "SET SCHEMA " + JdbcIdentifiers.INSTANCE.doubleQuote(schema);
    }

    @Override
    protected Object resultValue(ResultSet rs, int index, int sqlType) {
        return unchecked(() -> JdbcExecutor.INSTANCE.defaultResultValue(rs, index, sqlType));
    }

    private static String buildUrl(ConnectParams params) {
        if (params.getHost().isBlank()) {
            return "jdbc:h2:" + params.getDatabase();
        }
        return "jdbc:h2:tcp://" + params.getHost() + ":" + params.getPort() + "/" + params.getDatabase();
    }

    private static String resolveSchema(String schema) {
        if ("PUBLIC".equalsIgnoreCase(schema) || "INFORMATION_SCHEMA".equalsIgnoreCase(schema)) {
            return schema.toUpperCase(Locale.ROOT);
        }
        return "PUBLIC";
    }

    private static Integer intOrNull(ResultSet rs, String column) throws Exception {
        Object value = rs.getObject(column);
        return value instanceof Number ? ((Number) value).intValue() : null;
    }

    public static void main(String[] args) {
        new JsonRpcServer(new H2Agent()).run();
    }
}

File: h2/src/main/kotlin/com/dbx/agent/h2/H2Agent.kt
package com.dbx.agent.h2

import com.dbx.agent.*
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.Types

class H2Agent : DatabaseAgent {

    private var connection: Connection? = null
    private var databaseName: String = ""

    override fun getConnection(): Connection? = connection

    companion object {
        private const val MAX_ROWS = 10000
        private val QUERY_PREFIXES = listOf("SELECT", "WITH", "SHOW", "DESCRIBE", "EXPLAIN")
    }

    override fun connect(params: ConnectParams) {
        Class.forName("org.h2.Driver")
        val url = if (params.host.isBlank()) {
            "jdbc:h2:${params.database}"
        } else {
            "jdbc:h2:tcp://${params.host}:${params.port}/${params.database}"
        }
        connection = DriverManager.getConnection(url, params.username, params.password)
        databaseName = params.database
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName("org.h2.Driver")
        val url = if (params.host.isBlank()) {
            "jdbc:h2:${params.database}"
        } else {
            "jdbc:h2:tcp://${params.host}:${params.port}/${params.database}"
        }
        DriverManager.getConnection(url, params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        return listOf(DatabaseInfo(databaseName.ifBlank { "default" }))
    }

    override fun listSchemas(): List<String> {
        val conn = requireConnection()
        conn.prepareStatement(
            "SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA ORDER BY SCHEMA_NAME"
        ).use { stmt ->
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(rs.getString(1))
                    }
                }
            }
        }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = requireConnection()
        val effectiveSchema = resolveSchema(schema)
        conn.prepareStatement(
            """
            SELECT TABLE_NAME, TABLE_TYPE
            FROM INFORMATION_SCHEMA.TABLES
            WHERE TABLE_SCHEMA = ?
            ORDER BY TABLE_NAME
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val tableType = rs.getString("TABLE_TYPE").let {
                            if (it == "BASE TABLE") "TABLE" else it
                        }
                        add(TableInfo(rs.getString("TABLE_NAME"), tableType))
                    }
                }
            }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = requireConnection()
        val effectiveSchema = resolveSchema(schema)

        // Get primary key columns
        val primaryKeys = mutableSetOf<String>()
        conn.prepareStatement(
            """
            SELECT ic.COLUMN_NAME
            FROM INFORMATION_SCHEMA.INDEX_COLUMNS ic
            JOIN INFORMATION_SCHEMA.INDEXES i
              ON ic.INDEX_SCHEMA = i.INDEX_SCHEMA AND ic.INDEX_NAME = i.INDEX_NAME
            WHERE ic.TABLE_SCHEMA = ? AND ic.TABLE_NAME = ?
              AND i.INDEX_TYPE_NAME = 'PRIMARY KEY'
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                while (rs.next()) {
                    primaryKeys.add(rs.getString("COLUMN_NAME"))
                }
            }
        }

        // Get columns
        conn.prepareStatement(
            """
            SELECT COLUMN_NAME, DATA_TYPE, IS_NULLABLE, COLUMN_DEFAULT,
                   NUMERIC_PRECISION, NUMERIC_SCALE, CHARACTER_MAXIMUM_LENGTH
            FROM INFORMATION_SCHEMA.COLUMNS
            WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
            ORDER BY ORDINAL_POSITION
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val colName = rs.getString("COLUMN_NAME")
                        add(ColumnInfo(
                            name = colName,
                            data_type = rs.getString("DATA_TYPE"),
                            is_nullable = rs.getString("IS_NULLABLE") == "YES",
                            column_default = rs.getString("COLUMN_DEFAULT"),
                            is_primary_key = colName in primaryKeys,
                            extra = null,
                            comment = null,
                            numeric_precision = rs.getObject("NUMERIC_PRECISION")?.let { (it as Number).toInt() },
                            numeric_scale = rs.getObject("NUMERIC_SCALE")?.let { (it as Number).toInt() },
                            character_maximum_length = rs.getObject("CHARACTER_MAXIMUM_LENGTH")?.let { (it as Number).toInt() }
                        ))
                    }
                }
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        val conn = requireConnection()
        val effectiveSchema = resolveSchema(schema)
        conn.prepareStatement(
            """
            SELECT i.INDEX_NAME, ic.COLUMN_NAME, ic.IS_UNIQUE, i.INDEX_TYPE_NAME
            FROM INFORMATION_SCHEMA.INDEX_COLUMNS ic
            JOIN INFORMATION_SCHEMA.INDEXES i
              ON ic.INDEX_SCHEMA = i.INDEX_SCHEMA AND ic.INDEX_NAME = i.INDEX_NAME
            WHERE ic.TABLE_SCHEMA = ? AND ic.TABLE_NAME = ?
            ORDER BY i.INDEX_NAME, ic.ORDINAL_POSITION
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                val indexMap = linkedMapOf<String, MutableList<String>>()
                val uniqueMap = mutableMapOf<String, Boolean>()
                val primaryMap = mutableMapOf<String, Boolean>()
                val typeMap = mutableMapOf<String, String>()

                while (rs.next()) {
                    val indexName = rs.getString("INDEX_NAME")
                    val colName = rs.getString("COLUMN_NAME")
                    val isUnique = rs.getBoolean("IS_UNIQUE")
                    val indexType = rs.getString("INDEX_TYPE_NAME")

                    indexMap.getOrPut(indexName) { mutableListOf() }.add(colName)
                    uniqueMap[indexName] = isUnique
                    primaryMap[indexName] = indexType == "PRIMARY KEY"
                    typeMap[indexName] = indexType ?: ""
                }

                return indexMap.map { (name, cols) ->
                    IndexInfo(
                        name = name,
                        columns = cols,
                        is_unique = uniqueMap[name] ?: false,
                        is_primary = primaryMap[name] ?: false,
                        index_type = typeMap[name]
                    )
                }
            }
        }
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        val conn = requireConnection()
        val effectiveSchema = resolveSchema(schema)
        conn.prepareStatement(
            """
            SELECT FK_NAME, FKCOLUMN_NAME, PKTABLE_NAME, PKCOLUMN_NAME
            FROM INFORMATION_SCHEMA.CROSS_REFERENCES
            WHERE FKTABLE_SCHEMA = ? AND FKTABLE_NAME = ?
            ORDER BY FK_NAME
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(ForeignKeyInfo(
                            name = rs.getString("FK_NAME"),
                            column = rs.getString("FKCOLUMN_NAME"),
                            ref_table = rs.getString("PKTABLE_NAME"),
                            ref_column = rs.getString("PKCOLUMN_NAME")
                        ))
                    }
                }
            }
        }
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        val conn = requireConnection()
        val effectiveSchema = resolveSchema(schema)
        conn.prepareStatement(
            """
            SELECT TRIGGER_NAME, EVENT_MANIPULATION, ACTION_TIMING
            FROM INFORMATION_SCHEMA.TRIGGERS
            WHERE TRIGGER_SCHEMA = ? AND EVENT_OBJECT_TABLE = ?
            ORDER BY TRIGGER_NAME
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, effectiveSchema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(TriggerInfo(
                            name = rs.getString("TRIGGER_NAME"),
                            event = rs.getString("EVENT_MANIPULATION"),
                            timing = rs.getString("ACTION_TIMING")
                        ))
                    }
                }
            }
        }
    }

    override fun executeQuery(sql: String, schema: String?): QueryResult {
        val conn = requireConnection()
        val trimmedSql = sql.trim().trimEnd(';')
        val upperSql = trimmedSql.uppercase().trimStart()

        // Transaction control
        if (upperSql == "BEGIN" || upperSql == "BEGIN TRANSACTION") {
            val start = System.currentTimeMillis()
            conn.autoCommit = false
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }
        if (upperSql == "COMMIT") {
            val start = System.currentTimeMillis()
            conn.commit()
            conn.autoCommit = true
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }
        if (upperSql == "ROLLBACK") {
            val start = System.currentTimeMillis()
            conn.rollback()
            conn.autoCommit = true
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }

        // Set schema if provided
        if (!schema.isNullOrBlank()) {
            conn.createStatement().use { stmt ->
                stmt.execute(setSchemaSQL(schema))
            }
        }

        val startTime = System.currentTimeMillis()
        val isQuery = QUERY_PREFIXES.any { upperSql.startsWith(it) }

        return if (isQuery) {
            conn.createStatement().use { stmt ->
                stmt.maxRows = MAX_ROWS + 1
                stmt.executeQuery(trimmedSql).use { rs ->
                    val meta = rs.metaData
                    val colCount = meta.columnCount
                    val columns = (1..colCount).map { meta.getColumnLabel(it) }
                    val rows = mutableListOf<List<Any?>>()

                    while (rs.next() && rows.size < MAX_ROWS) {
                        val row = (1..colCount).map { i ->
                            val value = rs.getObject(i)
                            if (rs.wasNull()) null else value?.toString()
                        }
                        rows.add(row)
                    }

                    val truncated = rs.next()
                    val elapsed = System.currentTimeMillis() - startTime

                    QueryResult(
                        columns = columns,
                        rows = rows,
                        affected_rows = 0,
                        execution_time_ms = elapsed,
                        truncated = truncated
                    )
                }
            }
        } else {
            conn.createStatement().use { stmt ->
                val affected = stmt.executeUpdate(trimmedSql)
                val elapsed = System.currentTimeMillis() - startTime
                QueryResult(
                    columns = emptyList(),
                    rows = emptyList(),
                    affected_rows = affected.toLong(),
                    execution_time_ms = elapsed
                )
            }
        }
    }

    override fun setSchemaSQL(schema: String): String = "SET SCHEMA \"$schema\""

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }

    private fun resolveSchema(schema: String): String {
        if (schema.equals("PUBLIC", ignoreCase = true) || schema.equals("INFORMATION_SCHEMA", ignoreCase = true)) {
            return schema.uppercase()
        }
        return "PUBLIC"
    }
}

fun main() {
    JsonRpcServer(H2Agent()).run()
}
